README
=== README for Google-Merchant version 0.15
=   Generated on Sun Jan  5 19:04:10 2014 by OODoc 2.01

There are various ways to install this module:

 (1) if you have a command-line, you can do:
       perl -MCPAN -e 'install <any package from this distribution>'

 (2) if you use Windows, have a look at http://ppm.activestate.com/

 (3) if you have downloaded this module manually (as root/administrator)
       gzip -d Google-Merchant-0.15.tar.gz
       tar -xf Google-Merchant-0.15.tar
       cd Google-Merchant-0.15
       perl Makefile.PL
       make          # optional
       make test     # optional
       make install

For usage, see the included manual-pages or
    http://search.cpan.org/dist/Google-Merchant-0.15/

Please report problems to
    http://rt.cpan.org/Dist/Display.html?Queue=Google-Merchant



ChangeLog
====== version history of Google::Merchant

version 0.15: Sun Jan  5 19:03:52 CET 2014
	Improvements:
	- use ::Cache::addPrefixes()
	- add google_product_category field

version 0.14: Tue May 28 12:20:35 CEST 2013
	- AtomFeed: entry description is optional, but results in
	  warnings in Google's upload report.
	- Google gets confused about atom attribute "type".  Remove them.

version 0.13: Tue May 21 10:15:02 CEST 2013
	- Google's server does not handle namespaces correctly: atom must
	  be default and base elements must use 'g:'.
	- xmlns:g missing in result
	- added more fields to xsd

version 0.12: Fri May 17 11:55:51 CEST 2013
	- minor documentation cleanups
	- pod-tail not related to XML::Compile -- now for real ;-)
	- many fixes to the google-base schema...
	- latin1/utf8 fixes

version 0.11: Mon May 13 16:19:17 CEST 2013
	- make XML as slave object, not base-class
	- support custom extensions.
	- pod-tail not related to XML::Compile

version 0.10: Mon May 13 12:17:23 CEST 2013
	- initial release


Makefile.PL
use ExtUtils::MakeMaker;

use 5.008;

WriteMakefile
 ( NAME	 	=> 'Google::Merchant'
 , VERSION	=> '0.15'
 , PREREQ_PM    =>
   { XML::Compile::Cache       => 0.995
   , XML::Compile              => 1.34
   , XML::Compile::Tester      => 0
   , Log::Report               => 0.19
   , Test::More                => 0.54
   }

 , EXE_FILES    => []
 , AUTHOR       => 'Mark Overmeer'
 , ABSTRACT     => 'Google Merchant interface'
 , LICENSE      => 'perl'
 );

### used by oodist during production of distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
EXTENDS		= ../XMLCompile:../XMLCache
RAWDIR          = ../public_html/google-merchant/raw
DISTDIR         = ../public_html/google-merchant/source
LICENSE         = artistic
SKIP_LINKS	= XML::LibXML

# for POD
FIRST_YEAR      = 2013
EMAIL           = perl@overmeer.net
WEBSITE         = http://perl.overmeer.net/xml-compile/
__POSTAMBLE


lib/Google/Merchant.pm
# Copyrights 2013-2014 by [Mark Overmeer].
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 2.01.
use warnings;
use strict;

package Google::Merchant;
use vars '$VERSION';
$VERSION = '0.15';


use Log::Report 'google-merchant';

use XML::Compile::Cache    ();
use XML::LibXML            ();
use Google::Merchant::Util qw/:ns10/;
use Scalar::Util           qw/blessed/;
use Encode                 qw/encode/;

my $schemas;


sub new($%) { my $class = shift; (bless {}, $class)->init( {@_} ) }

sub init($)
{   my ($self, $args) = @_;

    unless($schemas)
    {   $schemas = $self->_loadSchemas;
        $schemas->compileAll;
    }

    $self->{GM_feed}        = {};
    $self->{GM_string_format} = $args->{string_format} || 'HTML';
    $self;
}

sub _loadSchemas()
{   my $self = shift;

    $schemas = XML::Compile::Cache->new
      ( # prefixes must be kept short, to reduce transported file size
        prefixes    => [g => NS_GOOGLE_BASE10, c => NS_GOOGLE_CUSTOM10]
      , any_element => 'TAKE_ALL'
      );
    $self->_loadXSD($schemas, 'google-base-10.xsd')
         ->_loadXSD($schemas, 'google-base-10-bug.xsd');

    $schemas->declare(WRITER => 'g:item', hooks =>
      [ { type    => 'g:stringAttrValueType'
        , replace => sub {$self->_write_string(@_)} }
      ] );

    $schemas;
}

sub _loadXSD($)
{   my ($self, $schemas, $schema_fn) = @_;
    (my $fn = __FILE__) =~ s!([/\\])(\w+)\.pm$!$1$2$1xsd$1$schema_fn!;
    $schemas->importDefinitions($fn);
    $self;
}

#---------

sub feed()    {shift->{GM_feed}}
sub schemas() {$schemas}

#---------

sub stringFormat() { shift->{GM_string_format} }


sub addItem() {panic "not implemented"}

sub _baseItem($)
{   my ($self, $google) = @_;
    
    # more smart behavior may be required for the google base params
    $google;
}

#----------------

sub write($%)
{   my ($self, $fn, %args) = @_;
    $args{doc}  ||= XML::LibXML::Document->new('1.0', 'UTF-8');
    $args{feed}   = $self->feed;
    $self->_write($fn, \%args);
}

sub _write($$) { panic "not implemented" }

sub _write_base_entry($$)
{   my ($self, $doc, $base) = @_;
    $self->schemas->writer('g:item')->($doc, $base);
}

sub _write_string($$$$$)
{   my ($self, $doc, $val, $path, $tag, $r) = @_;
    return $val if blessed $val && $val->isa('XML::LibXML::Element');

    my $type;
    if(ref $val eq 'HASH')
    {   $type = $val->{type};
        $val  = $val->{_};
    }
    else
    {   $type = $self->stringFormat;
    }

    my $str = encode 'utf8', $val;
    $str = XML::LibXML::CDATASection->new($str)
        if $type ne 'TEXT' && $str =~ m/\&/;

    $r->($doc, $str);
}

#------------

1;


lib/Google/Merchant/AtomFeed.pm
# Copyrights 2013-2014 by [Mark Overmeer].
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 2.01.
use warnings;
use strict;

package Google::Merchant::AtomFeed;
use vars '$VERSION';
$VERSION = '0.15';

use base 'Google::Merchant';

use Log::Report 'google-merchant';

use Google::Merchant::Util qw/:ns10/;
use XML::Compile::Util     qw/XMLNS/;
use Encode                 qw/encode/;


sub init($)
{   my ($self, $args) = @_;
    $self->SUPER::init($args);

    my $title  = $args->{title}   || panic "feed title required";
    my $site   = $args->{website} || panic "feed website required";

    # Google only supports these three, it seems
    my @globs;
    push @globs, +{title   => $title};
    push @globs, +{link    => +{rel => 'self', href => $site}};
    push @globs, +{updated => ($args->{updated} || time) };

    {  no strict;  # in devel, $VERSION undeclared
       push @globs, +{generator => __PACKAGE__ . ' '. ($VERSION || 'undef')}
    }
    push @globs, +{entry => ($self->{GMA_entries} = []) };

    my $feed = $self->feed;
    $feed->{cho_author} = \@globs;
    $feed->{base} = $args->{base} || $site;
    $feed->{lang} = $args->{language};

    $self;
}

sub _loadSchemas()
{   my $self = shift;
    my $schemas = $self->SUPER::_loadSchemas();

    $schemas->addPrefixes('' => NS_ATOM_2005);
    $self->_loadXSD($schemas, 'atom-2005.xsd');
    $schemas->importDefinitions(XMLNS);
    $schemas->declare
      ( WRITER => 'feed'
      , mixed_elements => 'STRUCTURAL'
      , hook   =>
         [ +{ type    => 'textType'
            , replace => sub {$self->_write_texttype(@_) }
            }
         , +{ type    => 'entryType'
            , replace => sub {$self->_write_entrytype(@_)}
            }
         ]
      );

    $schemas;
}

#---------

sub addItem(%)
{   my ($self, %args) = @_;

    my $title = delete $args{title}       or panic "entry title required";
    my $page  = delete $args{webpage}     or panic "entry webpage required";
    my $descr = delete $args{description};

    my @glob;
    push @glob, +{title  => $title};
    push @glob, +{link   => +{href => $page}};
    push @glob, +{summary => $descr};
    my $entry = +{cho_author => \@glob, _base => $self->_baseItem(\%args)};

    push @{$self->{GMA_entries}}, $entry;
    $entry;
}

#----------------

sub _write($$)
{   my ($self, $fn, $args) = @_;
    my $format = $args->{beautify} || 0;
    my $doc    = $args->{doc};

    my $root   = $self->schemas->writer('feed')->($doc, $args->{feed});
    $root->setNamespace(NS_GOOGLE_BASE10,   'g', 0);
    $root->setNamespace(NS_GOOGLE_CUSTOM10, 'c', 0);
    $doc->setDocumentElement($root);
    $doc->setCompression($args->{gzip}) if defined $args->{gzip};
    $doc->toFile($fn, $format);
    $self;
}

# Hook, because textType is mixed!
sub _write_texttype($$$$$)
{   my ($self, $doc, $val, $path, $tag, $r) = @_;
    my ($text, %attr);
    if(ref $val eq 'HASH')
    {   %attr = %$val;
        $text = delete $attr{_};
    }
    else
    {   $text = $val;
    }

    # Clean to use 'type' in atom elements, however the google server
    # confuses this  with its own "type" attributes.
#   $attr{type} ||= lc $self->stringFormat;

    my $xml  = $r->($doc, \%attr);
    $xml->appendText(encode 'utf8', $text);
    $xml;
}

sub _write_entrytype($$$$$)
{   my ($self, $doc, $val, $path, $tag, $r) = @_;
    my $base  = $self->_write_base_entry($doc, delete $val->{_base});
    my $node  = $r->($doc, $val);
    $node->appendChild($_) for $base->childNodes;

    # The xmlns:g is lost here, added globally
    $node;
}

1;
